README.md
# Kids First Snapshot Task Service
Snapshot Task Service is a standalone application running against Kids First Release Coordinator.  
Following the [Task Service Specfications][1], the service saves data dumps for releases to AWS S3.

## Getting Started (TBD)

## Operations
On a release, this service requests all entity endpoints for all studies included in the release.
Only entities with `visible=True` will be stored in a snapshot.
It tars JSONs from the Data Service endpoints and uploads to S3 in the following folder structure:
```
RE_00000001
    └── SD_00000001
        └── dataservice
                └── SD.json
                └── PT.json
                └── DG.json
                └── PH.json
                └── ...
        └── elasticsearch
               └── participant_centric_SD_00000001_RE_00000001.tar.gz
               └── file_centric_SD_00000001_RE_00000001.tar.gz
        └── ...
    └── SD_00000002
        └── ...
    └── RE_00000001.tar.gz  <- Everything above, tarred+compressed
└── ...
```

[1]: https://github.com/kids-first/kf-api-release-coordinator


CHANGELOG.md
# Kids First Task Snapshot Release 1.2.0

## Features

### Summary

Feature Emojis: 🐛x1 x1 ♻️x1
Feature Labels: [bug](https://api.github.com/repos/kids-first/kf-task-snapshot/labels/bug) x1 [feature](https://api.github.com/repos/kids-first/kf-task-snapshot/labels/feature) x1 [refactor](https://api.github.com/repos/kids-first/kf-task-snapshot/labels/refactor) x1

### New features and changes

- (#43) 🐛 Debug snapshot object storage - @liberaliscomputing
- (#40)  🎉 Handle file format query - @liberaliscomputing
- (#39) ♻️ Refactor single study download - @liberaliscomputing


# Kids First Task Snapshot Release 1.1.0

## Features

### Summary

Feature Emojis: 🎉Addx1 🚧x1
Feature Labels: [feature](https://api.github.com/repos/kids-first/kf-task-snapshot/labels/feature) x1 [bug](https://api.github.com/repos/kids-first/kf-task-snapshot/labels/bug) x1

### New features and changes

- (#34) 🎉Add download route - @liberaliscomputing
- (#28) 🚧 Debug staging failure - @liberaliscomputing


# Kids First Task Snapshot Release 1.0.0

## Features

### Summary

Feature Emojis: 🎉x4 🔧x3 📝x1
Feature Labels: [feature](https://api.github.com/repos/kids-first/kf-task-snapshot/labels/feature) x4 [refactor](https://api.github.com/repos/kids-first/kf-task-snapshot/labels/refactor) x2 [documentation](https://api.github.com/repos/kids-first/kf-task-snapshot/labels/documentation) x1 [ready-for-review](https://api.github.com/repos/kids-first/kf-task-snapshot/labels/ready-for-review) x1 [bug](https://api.github.com/repos/kids-first/kf-task-snapshot/labels/bug) x1

### New features and changes

- (#26) 🎉 Reorganize snapshot structure - @liberaliscomputing
- (#25) 🎉 Snapshot visible entities - @liberaliscomputing
- (#24) 🔧 Modify linting - @liberaliscomputing
- (#22) 📝 Update folder structure - @dankolbman
- (#18) 🎉 Add DynamoDB for persistent task_id management - @liberaliscomputing
- (#17) 🔧 Add application/json meme type - @liberaliscomputing
- (#13) 🔧 Fix #12 - @liberaliscomputing
- (#10) 🎉 Add Dockerfile, Jenkinsfile, and .circlrci and Refactor in Progress - @liberaliscomputing


test/test.spec.js
const request = require('request');
const { expect } = require('chai');

const app = require('../app');


const host = process.env.HOST || 'localhost';
const port = process.env.PORT || 3030;
let server;

before(() => { server = app.listen(port); });

const options = {
  baseUrl: `http://${host}:${port}`,
  headers: {
    'Content-Type': 'application/json',
  },
};

describe('Sending a request', () => {
  let data;

  describe('GET /status', () => {
    options.uri = '/status';
    options.method = 'GET';

    it('should return statusCode equal to 200', () => {
      request(options, (err, res) => {
        expect(res.statusCode).to.equal(200);
      });
    });

    data = JSON.stringify({
      message: {
        name: 'snapshot task',
        version: '1.2.0',
      },
    });
    it(`should return body equal to ${data}`, () => {
      request(options, (err, res, body) => {
        expect(body).to.equal(data);
      });
    });
  });

  describe('GET /download', () => {
    options.uri = '/download';

    it('should return statusCode equal to 400', () => {
      request(options, (err, res) => {
        expect(res.statusCode).to.equal(400);
      });
    });

    data = JSON.stringify({
      message: 'missing a required field',
    });
    it(`should return body equal to ${data}`, () => {
      request(options, (err, res, body) => {
        expect(body).to.equal(data);
      });
    });

    options.uri += '/RE_00000001';

    it('should return statusCode equal to 404', () => {
      request(options, (err, res) => {
        expect(res.statusCode).to.equal(404);
      });
    });

    data = JSON.stringify({
      message: 'release not found',
    });
    it(`should return body equal to ${data}`, () => {
      request(options, (err, res, body) => {
        expect(body).to.equal(data);
      });
    });

    options.uri = '/download/RE_5N9ZMDSM/SD_00000001';

    it('should return statusCode equal to 404', () => {
      request(options, (err, res) => {
        expect(res.statusCode).to.equal(404);
      });
    });

    data = JSON.stringify({
      message: 'study not found',
    });
    it(`should return body equal to ${data}`, () => {
      request(options, (err, res, body) => {
        expect(body).to.equal(data);
      });
    });
  });
});

after(() => { server.close(); });


routes/download.js
const { Readable } = require('stream');

const express = require('express');
const AWS = require('aws-sdk');
const zlib = require('zlib');


AWS.config.update({ region: 'us-east-1' });
const s3 = new AWS.S3();

const router = express.Router();
require('dotenv').config();

// handles when missing release_id
router.get('/', (req, res) => {
  res.status(400).json({
    message: 'missing a required field',
  });
});

// handles single release/study download
router.get('/:release_id/:study_id?', (req, res) => {
  const { release_id, study_id } = req.params;
  let { file_format } = req.query;

  // validate file_format enums
  if (file_format === undefined) file_format = 'gz';
  else {
    file_format = file_format.toLowerCase();
    if (['gz', 'zip'].indexOf(file_format) === -1) {
      return res.status(400).json({
        message: 'file_format must be one of: gz, zip',
      });
    }
  }

  let Prefix;
  if (!study_id) Prefix = `${release_id}`;
  else Prefix = `${release_id}/${study_id}`;

  const params = { Bucket: SNAPSHOT_BUCKET, Prefix };

  listObjectsAsync(params)
    .then((body) => {
      // handles if a release/study does not exist
      if (body.Contents.length === 0) {
        res.status(404).json({
          message: `${study_id || release_id} not found`,
        });
      }

      return getObjectsAsync(params, body.Contents);
    })
    .then((obj) => {
      // set response headers
      res.attachment(
        `${study_id || release_id}.json.${file_format}`,
      );

      // create a readable object stream
      const readStream = new Readable();
      readStream.push(JSON.stringify(obj));
      readStream.push(null);

      // pipe readStream into gzip
      const gzip = zlib.createGzip();
      readStream.pipe(gzip).pipe(res);
    })
    .catch((err) => {
      return res.status(err.statusCode).json(err);
    });
});

// import environment variables
const { SNAPSHOT_BUCKET } = process.env;

/**
 * @access private
 * listObjectsAsync() lists objects in S3
 * @param {Object} params
 * @returns {Object}
 */
const listObjectsAsync = (params) => {
  return new Promise((resolve, reject) => {
    const { Bucket, Prefix } = params;
    console.log(`LISTing ${Prefix} in ${Bucket}`);
    s3.listObjects(params, (err, body) => {
      err ? reject(err) : resolve(body);
    });
  });
};

/**
 * @access private
 * getObjectsAsync() retrieves objects in S3
 * @param {Object} params
 * @param {Object} contents
 * @returns {Object}
 */
const getObjectsAsync = (params, contents) => {
  return new Promise((resolve, reject) => {
    const obj = {};
    delete params.Prefix;
    let count = contents.length;
    contents.forEach((content) => {
      const { Key } = content;
      params.Key = Key;

      getObjectAsync(params)
        .then((body) => {
          nest(obj, Key.split('/'), JSON.parse(body.Body));
          count -= 1;
          if (count <= 0) resolve(obj);
        })
        .catch((err) => { reject(err); });
    });
  });
};

/**
 * @access private
 * getObjectAsync() retrieves an object in S3
 * @param {Object} params
 * @returns {Object}
 */
const getObjectAsync = (params) => {
  return new Promise((resolve, reject) => {
    const { Bucket, Key } = params;
    console.log(`GETting ${Key} in ${Bucket}`);
    s3.getObject(params, (err, body) => {
      err ? reject(err) : resolve(body);
    });
  });
};

/**
 * @access private
 * nest() creates a nested object
 * @param {Object} obj
 * @param {Object[]} keys
 * @param {Object} value
 * @returns {Object}
 */
const nest = (obj, keys, value) => {
  const key = keys.shift().split('.').shift();
  if (keys.length === 0) obj[key] = value;
  else {
    obj[key] = nest(
      obj[key] === undefined ? {} : obj[key], keys, value,
    );
  }

  return obj;
};

module.exports = router;


routes/tasks.js
const express = require('express');
const request = require('request');
const AWS = require('aws-sdk');

const ENDPOINTS = require('../config/endpoint');


AWS.config.update({ region: 'us-east-1' });
const docClient = new AWS.DynamoDB.DocumentClient();
const s3 = new AWS.S3();

const router = express.Router();
require('dotenv').config();

router.post('/', async (req, res) => {
  const { action, task_id, release_id } = req.body;

  // if any of the params misses, return 400
  if (!action || !task_id || !release_id) {
    return res.status(400).json({
      message: 'missing a required field',
    });
  }

  // if an action is not pre-defined, return 400
  if (ACTIONS[action] === undefined) {
    const actions = Object.keys(ACTIONS).join(', ');
    return res.status(400).json({
      message: `action must be one of: ${actions}`,
    });
  }

  // if a task is not initialized, return 400
  const task = await getTask(task_id);
  if (task === undefined && action !== 'initialize') {
    return res.status(400).json({
      message: `${task_id} not found. Initialize it`,
    });
  }

  // if an action is not allowed, return 400
  const state = task !== undefined ? task.state : task;
  let isAllowed = TRANSITIONS[state] === action;
  isAllowed = (isAllowed
    || ['get_status', 'cancel'].indexOf(action) > -1);
  if (isAllowed === false) {
    return res.status(400).json({
      message: `${action} not allowed in ${state}`,
    });
  }

  // initialize a task
  if (task === undefined && action === 'initialize') {
    await initialize(task_id, release_id);
    return res.status(200).json(await getTask(task_id));
  }

  // trigger start(), publish(), or cancel()
  if (['initialize', 'get_status'].indexOf(action) < 0) {
    const data = {};
    switch (action) {
      case 'start': {
        data.state = 'running';
        break;
      }
      case 'publish': {
        data.state = 'publishing';
        break;
      }
      case 'cancel': {
        data.state = 'cancelling';
        break;
      }
      // no default
    }

    await updateState(task_id, data);
    ACTIONS[action](task_id, release_id);
  }

  // return the status of task
  return res.status(200).json(await get_status(task_id));
});

// define a snapshot cache
const snapshot = {};

// define constants
const {
  DATASERVICE_API,
  COORDINATOR_API,
  TABLE_NAME,
  SNAPSHOT_BUCKET,
} = process.env;
const TRANSITIONS = {
  undefined: 'initialize',
  pending: 'start',
  staged: 'publish',
};
const REQUEST_OPTIONS = {
  release: {
    uri: '/releases',
    baseUrl: COORDINATOR_API,
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
  },
  data: {
    baseUrl: DATASERVICE_API,
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
    qs: { limit: 100, visible: true },
  },
  patch: {
    uri: '/tasks',
    baseUrl: COORDINATOR_API,
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: { progress: 100, state: null },
    json: true,
  },
};

/**
 * @access public
 * initialize() creates a new task of snapshot
 * @param {string} task_id
 * @param {string} release_id
 */
const initialize = (task_id, release_id) => {
  const data = {
    task_id,
    release_id,
    name: 'snapshot task',
    date_submitted: getISOString(new Date()),
    progress: null,
    state: 'pending',
  };

  // initialize a release snapshot
  snapshot[release_id] = {};

  // update task in DynamoDB to pending
  return updateState(task_id, data);
};

/**
 * @access private
 * getISOString() converts a Date object to an ISO string
 * @param {Object} date
 * @returns {string}
 */
const getISOString = (date) => {
  const checkDigit = (num) => {
    return (num < 10 ? '0' : '') + num;
  };

  return `${date.getUTCFullYear()}-`
    + `${checkDigit(date.getUTCMonth() + 1)}-`
    + `${checkDigit(date.getUTCDate())}T`
    + `${checkDigit(date.getUTCHours())}:`
    + `${checkDigit(date.getUTCMinutes())}:`
    + `${checkDigit(date.getUTCSeconds())}Z`;
};

/**
 * @access public
 * start() triggers the creation of snapshot
 * @param {string} task_id
 * @param {string} release_id
 * @returns {Object}
 */
const start = (task_id, release_id) => {
  const releaseOptions = { ...REQUEST_OPTIONS.release };
  releaseOptions.uri += `/${release_id}`;

  requestAsync(releaseOptions)
    .then(({ res, body }) => {
      // get study IDs in a given release
      return JSON.parse(body).studies;
    })
    .then((studies) => {
      // get and store entities by study
      return Promise.all(studies.map((study) => {
        return scrapeByStudy(REQUEST_OPTIONS.data, release_id, study);
      }));
    })
    .then((studies) => {
      // put objects in S3 by study and entity
      return Promise.all(studies.map((study) => {
        return createSnapshot(release_id, study);
      }));
    })
    .then((studies) => {
      // patch task in release coordinator to staged
      console.log(`STAGED: ${studies.join(', ')}`);

      const patchOptions = { ...REQUEST_OPTIONS.patch };
      patchOptions.uri += `/${task_id}`;
      patchOptions.body.state = 'staged';

      return requestAsync(patchOptions);
    })
    .then(() => {
      // update task in DynamoDB to staged
      updateState(task_id, { state: 'staged' });
    })
    .catch((err) => {
      // update task in DynamoDB to failed
      updateState(task_id, { state: 'failed' });
      console.log(err.message);
    })
    .finally(() => {
      // flush out when staged or failed
      delete snapshot[release_id];
    });
};

/**
 * @access private
 * requestAsync() sends a promisified request
 * @param {Object} options
 * @returns {Object}
 */
const requestAsync = (options) => {
  return new Promise((resolve, reject) => {
    request(options, (err, res, body) => {
      if (err) reject(err);
      else resolve({ res, body });
    });
  });
};

/**
 * @access private
 * loopRquest() paginates and saves the results
 * @param {Object} options
 * @param {string} link
 * @param {Object[]} array
 * @returns {Object}
 */
const loopRequest = (options, link, array) => {
  options.uri = link;
  return new Promise((resolve, reject) => {
    requestAsync(options)
      .then(({ res, body }) => {
        const data = JSON.parse(body);
        // handles a 404 response
        if (data._status.code === 404) resolve(array);

        let { results } = data;
        results = array.concat(results);
        const { next } = data._links;
        if (next === undefined) resolve(results);
        else resolve(loopRequest(options, next, results));
      })
      .catch((err) => { reject(err); });
  });
};

/**
 * @access private
 * scrapeByStudy() collects data by study
 * @param {Object} options
 * @param {string} release_id
 * @param {string} study
 * @returns {string}
 */
const scrapeByStudy = (options, release_id, study) => {
  return new Promise((resolve, reject) => {
    console.log(`START: scraping ${study} in ${release_id}`);

    snapshot[release_id][study] = {};
    const entries = Object.entries(ENDPOINTS);
    let count = entries.length;
    entries.forEach((entry) => {
      const entity = entry[0];
      let endpoint = entry[1];
      if (entity === 'study') endpoint += `/${study}`;
      else options.qs.study_id = study;

      loopRequest(options, endpoint, [])
        .then((results) => {
          snapshot[release_id][study][entity] = results;
          const size = snapshot[release_id][study][entity].length;
          console.log(`DONE: ${size} ${entity} of ${study}`);
          count -= 1;
          if (count <= 0) resolve(study);
        })
        .catch((err) => { reject(err); });
    });
  });
};

/**
 * @access private
 * createSnapshot() creates a snapshot by study and entity
 * @param {string} release_id
 * @param {string} study
 * @returns {Object}
 */
const createSnapshot = (release_id, study) => {
  return new Promise((resolve, reject) => {
    console.log(`START: creating ${study} in ${release_id}`);

    const entries = Object.entries(snapshot[release_id][study]);
    let count = entries.length;
    entries.forEach((entry) => {
      const [entity, data] = entry;
      putObjectAsync({
        Bucket: SNAPSHOT_BUCKET,
        Key: `${release_id}/${study}/dataservice/${entity}.json`,
        Body: JSON.stringify(data),
        ContentType: 'application/json',
      })
        .then((body) => {
          console.log(`DONE: ${body}`);
          count -= 1;
          if (count <= 0) resolve(study);
        })
        .catch((err) => { reject(err); });
    });
  });
};

/**
 * @access private
 * putObjectAsync() put data in an S3 bucket
 * @param {Object} params
 * @returns {Object}
 */
const putObjectAsync = (params) => {
  return new Promise((resolve, reject) => {
    console.log(`putting ${params.Key} into ${params.Bucket}`);
    s3.putObject(params, (err, body) => {
      if (err) reject(err);
      else resolve(body);
    });
  });
};

/**
 * @access public
 * publish() publishes a staged snapshot task
 * @param {string} task_id
 * @param {string} release_id
 * @returns {Object}
 */
const publish = (task_id, release_id) => {
  const releaseOptions = { ...REQUEST_OPTIONS.release };
  releaseOptions.uri += `/${release_id}`;

  requestAsync(releaseOptions)
    .then(({ res, body }) => {
      // get study IDs in a given release
      return JSON.parse(body).studies;
    })
    .then((studies) => {
      // TODO: add a proper publishing function
      return next(studies);
    })
    .then((studies) => {
      // patch task in release coordinator to published
      console.log(`PUBLISHED: ${studies.join(', ')}`);

      const patchOptions = { ...REQUEST_OPTIONS.patch };
      patchOptions.uri += `/${task_id}`;
      patchOptions.body.state = 'published';

      return requestAsync(patchOptions);
    })
    .then(() => {
      // update task in DynamoDB to published
      updateState(task_id, { state: 'published' });
    })
    .catch((err) => {
      // update task in DynamoDB to failed
      updateState(task_id, { state: 'failed' });
      console.log(err.message);
    });
};

/**
 * @access private
 * next() is a placeholding function
 * @param {Object[]} array
 * @returns {Object}
 */
const next = (array) => { return array; };

/**
 * @access public
 * cancel() terminates a snapshot task
 * @param {string} task_id
 * @param {string} release_id
 * @returns {Object}
 */
const cancel = async (task_id, release_id) => {
  // flush out when cancelled
  delete snapshot[release_id];
  await updateState(task_id, { state: 'cancelled' });
};

/**
 * @access public
 * get_status() returns a status of snapshot task
 * @param {string} task_id
 * @returns {Object}
 */
const get_status = (task_id) => { return getTask(task_id); };

/**
 * @access private
 * getTask() returns snapshot task information
 * @param {string} task_id
 * @returns {Object}
 */
const getTask = (task_id) => {
  const params = {
    TableName: TABLE_NAME, Key: { task_id },
  };

  return new Promise((resolve, reject) => {
    docClient.get(params, (err, body) => {
      if (err) reject(err);
      else resolve(body.Item);
    });
  });
};

/**
 * @access private
 * updateState() updates a state of snapshot task
 * @param {string} task_id
 * @param {Object} data
 */
const updateState = async (task_id, data) => {
  const task = await getTask(task_id);
  const params = { TableName: TABLE_NAME };

  if (!task) {
    return new Promise((resolve, reject) => {
      params.Item = data;
      docClient.put(params, (err, body) => {
        if (err) reject(err);
        else resolve(body);
      });
    });
  }

  return new Promise((resolve, reject) => {
    params.Key = { task_id };
    params.UpdateExpression = 'set #state = :s';
    params.ExpressionAttributeNames = {
      '#state': 'state',
    };
    params.ExpressionAttributeValues = {
      ':s': data.state,
    };
    docClient.update(params, (err, body) => {
      if (err) reject(err);
      else resolve(body);
    });
  });
};

// routing logic
const ACTIONS = {
  initialize, start, publish, get_status, cancel,
};

module.exports = router;
